src/main/java/frc/lib/util/AprilTag.java

```diff
@@ -173,4 +173,83 @@ public Optional<double[]> determineTargetRotationalOffset(Optional<Translation3d
 
         return Optional.of(new double[] {xAngularOffset, yAngularOffset});
     }
+
+    /**
+     * Returns the height difference between the robot and the target, if a target
+     * is detected.
+     * 
+     * @return an OptionalDouble representing the absolute height difference to the
+     *         target,
+     *         or an empty OptionalDouble if no target is detected.
+     */
+    public Optional<Double> getHeightDifferenceToTarget() {
+      // still need to implement the offsets between the actual target and the
+      // apriltag
+      if (targetPos().isPresent()) {
+        double heightDifference = targetPos().get().getY();
+        return Optional.of(Math.abs(heightDifference));
+      } else {
+        return Optional.empty();
+      }
+    }
+
+    /**
+     * Returns the vertical velocity component needed for the note to reach the
+     * target, if a
+     * target is detected.
+     * 
+     * @return an OptionalDouble representing the vertical velocity of the target,
+     *         or empty if no target is detected
+     */
+    public Optional<Double> getVerticalVelocity() {
+      if (targetPos().isPresent()) {
+        return Optional.of(Math.sqrt(
+            2 * Constants.Launcher.gravityAcceleration * getHeightDifferenceToTarget().get()));
+
+      } else {
+        return Optional.empty();
+      }
+    }
+
+    /**
+     * Calculates the time it takes for the note to travel to the target, if a
+     * target is detected.
+     * 
+     * @return An OptionalDouble representing the time to travel, or empty if there
+     *         is no target.
+     */
+    public Optional<Double> getTimeToTravel() {
+      if (targetPos().isPresent()) {
+        double timeToTravel = getVerticalVelocity().get() / Constants.Launcher.gravityAcceleration;
+        return Optional.of(timeToTravel);
+      }
+      return Optional.empty();
+    }
+
+
+    /**
+     * Returns the horizontal velocity component needed for the note to reach the
+     * target, if a
+     * target is detected.
+     * 
+     * @return an OptionalDouble representing the vertical velocity of the target,
+     *         or empty if no target is detected
+     */
+    public Optional<Double> getHorizontalVelocity() {
+      if (!targetPos().isPresent() && getHeightDifferenceToTarget().isPresent()) {
+        return Optional.empty();
+      }
+      //TODO: Use a new method to find the planar 2D distance to the target instead of getDirectDistance(), so that it disregards height.
```

Comment:
you can implement this by disregarding whichever axis is vertical and using something like Math.hypot